rpcs3/Emu/System.cpp

```diff
@@ -2585,257 +2583,311 @@ std::shared_ptr<utils::serial> Emulator::Kill(bool allow_autoexit, bool savestat
 	// Wait fot newly created cpu_thread to see that emulation has been stopped
 	id_manager::g_mutex.lock_unlock();
 
-	GetCallbacks().on_stop();
+	// Type-less smart pointer container for thread (cannot know its type with this approach)
+	// There is no race condition because it is only accessed by the same thread
+	std::shared_ptr<std::shared_ptr<void>> join_thread = std::make_shared<std::shared_ptr<void>>();
 
-	// Join threads
-	for (const auto& [type, data] : *g_fxo)
+	auto make_ptr = [](auto ptr)
 	{
-		if (type.stop)
+		return std::shared_ptr<std::remove_pointer_t<decltype(ptr)>>(ptr);
+	};
+
+	*join_thread = make_ptr(new named_thread("Emulation Join Thread"sv, [join_thread, savestate, allow_autoexit, this]() mutable
+	{
+		GetCallbacks().on_stop();
```

Comment:
Why is this called here?
The UI should only receive this signal after finish.
It probably wasn't an issue before since it was mostly blocking the mainthread anyway?

Reply:
I think it contains a hack for a debugger to invalidate its data internally at once without rsx thread destructor is called. Although it probably it can be removed now because of the added checks in #13964.

Reply:
why should it be removed? that makes no sense.
It should be moved to when we know the emulation fully stopped.